### PR TITLE
write_abstract_lef fixes

### PR DESCRIPTION
**Pull request by antonblanchard**
A few abstract LEF fixes, discovered when trying to replace the existing scripts in ORFS.

### CODE REVIEW

File: src/odb/src/lefout/lefout.cpp [hunk 1]
```diff
@@ -144,7 +158,34 @@ void lefout::writeObstructions(dbBlock* db_block)
...
+      Rectangle bounding_box_rect = Rectangle(block_bounding_box->xMin(), block_bounding_box->yMin(),
+                                               block_bounding_box->xMax(), block_bounding_box->yMax());
+      Polygon90Set obs_set = Polygon90Set(bounding_box_rect - bloat_set);
+
+      std::vector<Rectangle> rects;
+      obs_set.get_rectangles(rects);
+
+      for (auto& r : rects) {
+        writeBox("  ", xl(r), yl(r), xh(r), yh(r));
+      }
```
Comment (QuantamHD) [inline src/odb/src/lefout/lefout.cpp:185]: I think this code is great it's definitely more correct than what was here before, but will this actually impact the router?

From my understanding the router will override obstructions if the macro has pins on edge of the macro, and if the pin is on the inside of the macro it needs to accessible from either around the the pin or from above or below. In this case we're removing the obstructions around pins, but not from layers above or below.

File: src/odb/src/lefout/lefout.cpp [hunk 2]
```diff
@@ -144,7 +158,34 @@ void lefout::writeObstructions(dbBlock* db_block)
...
+        for (dbBPin* pin : term->getBPins()) {
+          for (dbBox* box : pin->getBoxes()) {
+            if (box->getTechLayer() == tech_layer) {
+              polygon_set.insert(Rectangle(box->xMin(), box->yMin(), box->xMax(), box->yMax()));
+            }
+          }
+        }
+      }
+
+      Polygon90Set bloat_set = Polygon90Set(polygon_set + tech_layer->getSpacing());
```
Comment (maliberty) [inline src/odb/src/lefout/lefout.cpp:175]: Min space is not necessarily correct.  A more than min-width pin shape would become a violation.

File: src/odb/src/lefout/lefout.cpp [hunk 3]
```diff
@@ -282,7 +328,7 @@ void lefout::writeBlock(dbBlock* db_block)
   fprintf(_out, "  ORIGIN %g %g ;\n", origin_x, origin_y);
   fprintf(_out, "  SIZE %g BY %g ;\n", size_x, size_y);
   writePins(db_block);
-  writeObstructions(db_block);
+  writeObstructions(db_block, true);
```
Comment (maliberty) [inline src/odb/src/lefout/lefout.cpp:331]: Should this be a user option rather than true?

File: src/odb/src/lefout/lefout.cpp (hunk 1 above)
Comment (antonblanchard) [reply src/odb/src/lefout/lefout.cpp:185]: > I think this code is great it's definitely more correct than what was here before, but will this actually impact the router?
> 
> From my understanding the router will override obstructions if the macro has pins on edge of the macro, and if the pin is on the inside of the macro it needs to accessible from either around the the pin or from above or below. In this case we're removing the obstructions around pins, but not from layers above or below.

Does grt and drt handle this? I made up a test case with a pin on the edge of a macro with an overlapping obstruction, and it failed in drt pin access.

File: src/odb/src/lefout/lefout.cpp (hunk 2 above)
Comment (antonblanchard) [reply src/odb/src/lefout/lefout.cpp:175]: @maliberty any thoughts on how best to handle this? Do we have to also look at the width of the obstruction next to the pin and lookup the spacing table based on the two widths, or can we get away with just using the widest clearance all the time?

I would guess obstructions between pins are almost always going to be big enough to invoke the widest spacing rule.

I'm also not sure if this direction (of creating clearance around pins) is what all the existing tools do and expect.

Comment (maliberty) [reply src/odb/src/lefout/lefout.cpp:175]: > I would guess obstructions between pins are almost always going to be big enough to invoke the widest spacing rule.
I think that is ok but you need to make sure there is actually no shape closer than what you are cutting away.

> I'm also not sure if this direction (of creating clearance around pins) is what all the existing tools do and expect.
I've seen it both ways - either with a blockage over the pins or a cutaway around them.  If the pins are not on the edge you must cut away the blockage but on the edge it isn't really a big difference.

File: src/odb/src/lefout/lefout.cpp (hunk 1 above)
Comment (QuantamHD) [reply src/odb/src/lefout/lefout.cpp:185]: I think what you have here is good, but if it doesn't work I believe that's a bug. @maliberty can you confirm.

Comment (maliberty) [reply src/odb/src/lefout/lefout.cpp:185]: The comments starting with "Min space is not necessarily correct" have not been addressed yet so it is not ready imo.

Comment (QuantamHD) [reply src/odb/src/lefout/lefout.cpp:185]: Sorry didn't mean to imply that this specific code should be merged, but the technique of removing obstructions around pins is good. I was asking you to confirm if Anton's case was in fact a bug.

Comment (maliberty) [reply src/odb/src/lefout/lefout.cpp:185]: I need to see the test case to see what the issue is. It should work.

**Comment by maliberty** [comment]
@antonblanchard are you still working on this PR?

**Comment by antonblanchard** [comment]
@maliberty I am, just got a bit side tracked with other bugs. Will get back to this one.

File: src/odb/src/lefout/lefout.cpp (hunk 1 above)
Comment (antonblanchard) [reply src/odb/src/lefout/lefout.cpp:185]: I'm not sure if it was recent updates to the pin access code, or just bugs in my test cases, but I've now tested three variants (macro sized obstruction, obstruction cut around pins with no clearance, clearance around pins). All work.

File: src/odb/src/lefout/lefout.cpp (hunk 3 above)
Comment (antonblanchard) [reply src/odb/src/lefout/lefout.cpp:331]: Do we want the option to enable, or disable clearance? My first attempt looks like:
```
+sta::define_cmd_args "write_abstract_lef" {[-pin_clearance] filename}
```

**Comment by maliberty** [comment]
Still relevant?

**Comment by maliberty** [comment]
@antonblanchard this PR is some old and needs conflict resolution.  If you'll update I can re-review

**Comment by maliberty** [comment]
Stale.  Reopen if you are still planning to work on it.

**Pull request closed without merging**